### DISCUSSION TITLE
Webhook question

### DISCUSSION THREAD

**Discussion by dejantep** [netbox-community/netbox#7574] [General]
I'm trying to learn how to use webhooks and i have created one as simple as possible, no body, basically just url and type create on dcim > site

When i create a site shouldn't i be able to see something under job results? i have nothing.
Dokumentation states following _webhook events are being placed into the queue (visible under the NetBox admin UI)_

**Comment by candlerb** [comment]
Look under "Background tasks".  "Job results" are for scripts and reports.

Otherwise: is the webhook enabled?  What Netbox version?  Do you have netbox-rq running?

It works for me.  Tested with Netbox 3.0.7:
* Other > Webhooks > +
* Name: foo
* Content types: DCIM > Site
* Type: Create (checked)
* URL: `http://localhost:1234/`
* Create

Now create a site:
* Organization > Sites > +
* Name: bar
* Create

Now look at webhook status:
* Admin > Admin
* Background tasks
  ![image](https://user-images.githubusercontent.com/44789/137942891-7613c394-0a78-40f1-b4c0-875409b4cfd9.png)
* Click on the `1` under Failed Jobs
  ![image](https://user-images.githubusercontent.com/44789/137943029-55cee536-6607-43d4-9f42-b51135247259.png)
* Click on the job ID
  ![image](https://user-images.githubusercontent.com/44789/137943209-3d64ccf4-bf16-4646-973e-a415b7b9871c.png)

**Comment by dejantep** [comment]
Thank you very much. Now i see what i'm missing. Menu choice "Background jobs" is only available when i'm on netbox administration home screen. If i go to webhooks, option is not there.

Another question...is secret field used to authenticate against target server or do i put that info in header section?

**Comment by candlerb** [comment]
It's documented [here](https://netbox.readthedocs.io/en/stable/additional-features/webhooks/#configuration):

> **Secret** - A secret string used to prove authenticity of the request (optional). This will append a `X-Hook-Signature` header to the request, consisting of a HMAC (SHA-512) hex digest of the request body using the secret as the key.

That is, in your webserver you can calculate a HMAC(secret key, request body) and check that it matches the X-Hook-Signature header.  If it doesn't match, the request is not authentic.  Here is some [sample code](https://github.com/candlerb/netbox-webhook-dnsupdate/blob/master/nb_dns_updater.py#L132-L141).

If you only want to provide a *static* password, then it can go in "additional headers".  In that case, you really want to make sure the connection is protected via TLS (i.e. the target provides a HTTPS URL, with a valid certificate)

**Reply by dejantep** [reply]
Great. Thanks for the help

**Comment by dejantep** [comment]
I would like to continue with another question instead of starting new thread.

One thing i dont get yet is accessing data context that is "two levels deep" for instance region of a site.
It is dict within data context and cannot be accessed like {{ data.region }}

How would i do that?

**Reply by candlerb** [reply]
The webhook [body template](https://netbox.readthedocs.io/en/stable/additional-features/webhooks/#jinja2-template-support) lets you control the data sent.  It can expand nested attributes.  The easiest place to play with this is is the [netbox shell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/) so you can test the attributes interactively.  (Jinja2 isn't python, but it's similar)

Alternatively, your webhook receiver can make a call back to the REST API (or the graphql API in 3.0) to get additional data.

**Reply by djgregor** [reply]
@dejantep I know this is an old thread, but I'm curious if you ever found a way to send the `region` field.  It sounds like our use case is similar to yours -- we're attempting to send `site.name` (address), `region` (city), and `site-group` (market/office, in our case).

Interestingly, we can send `data.site.region` and `data.site.group`, but the payload for both is empty.  Alternatively, if we can't achieve this cleanly with a body template, we're considering directing the webhook to trigger a local python script that could extract and forward all relevant data (definitely something we'd like to avoid though).  As a last resort, it looks like we could use custom fields for both region and site-group.

**Reply by dejantep** [reply]
Hello,
yes i found a way. Here is my webhook template. Hope it helps.

```
{
"name": "{{data.name}}",
"description": "{{data.description}}",
"placement": "{{data.facility}}",
"attributes": {
"Area": "{{data.region['name']}}",
"Street": "{{data.physical_address}}",
"NB-id": "{{data.id}}"
},
"longitude": "{{data.longitude}}",
"latitude": "{{data.latitude}}"
}
```

**Comment by dejantep** [comment]
I tried to start nbshell but with no success. Getting Django missing module error.
cd /opt/netbox/netbox
./manage.py nbshell

**Reply by candlerb** [reply]
```
cd /opt/netbox
. venv/bin/activate
python3 netbox/manage.py nbshell
```

**Reply by dejantep** [reply]
unfortunately not working yet. Here is the result:

(venv) dejan@netbox:/opt/netbox$ python3 netbox/manage.py nbshell
Traceback (most recent call last):
  File "netbox/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'

**Reply by candlerb** [reply]
Try "python" instead of "python3" (that should use the version of python linked from your virtualenv)

**Reply by dejantep** [reply]
i appreciate your help but i get the same error
inside venv i have python 3.7.5

**Reply by candlerb** [reply]
Something is apparently broken in your virtualenv, and it seems very odd that netbox works at all.

What OS is this? Where did you get that version of python?  Did you run `./upgrade.sh` as part of the installation process, and did it give any errors?

**Reply by dejantep** [reply]
I dont know if its broken but it has been working fine for quite some time. No issues during latest upgrade to netbox 2.11.12
Python version 3.7.5 is installed by me recently but i ddint give right info about my venv python. I believe this is right info:
dejan@netbox:~$ cat /opt/netbox/venv/pyvenv.cfg 
home = /usr/bin
include-system-site-packages = false
version = 3.6.9

OS is ubuntu 18.04.3 LTS

**Reply by candlerb** [reply]
This is just a guess, but have you symlinked `python3` to `python3.7` (either manually or by using update-alternatives?)

This is a really, really bad thing to do on Ubuntu.  It breaks all sorts of things, since many system utilities depend on the version of python originally supplied with Ubuntu.  I *strongly* recommend you put it back.  As you can see, your netbox venv has been built to run with python3.6 (which is fine for Netbox <3.0)

When you get around to upgrading to Netbox 3.0, then you'll do this:

```
PYTHON=python3.7 ./upgrade.sh
```

and that will tell the upgrade script to use that particular version of python when building the venv, even though `python3` still points at `python3.6`.  Alternatively, you may find it easier to take the opportunity to upgrade to Ubuntu 20.04 instead.

**Reply by chicks-net** [reply]
@candlerb I'm not saying it might be a bad idea, but Ubuntu provides `python-is-python2` and `python-is-python3` packages which create symlinks for `python` to one or the other.  I have been using `python-is-python3` on Ubuntu 20.04 LTS build agents without apparent breakage.

**Reply by candlerb** [reply]
python-is-python3 isn't a great idea either, but since most of the system utilities have been moved to python3 anyway that's not a major problem.

It's symlinking python3 to python3.7 which is worse (when the standard system python3 is python3.6)

**Reply by dejantep** [reply]
yes i have changed system default python3 version to 3.7.5 through "update-alternatives"
I'm not sure it broke anything but i understand your point. 
But if my venv is using 3.6.9 and venv purpose is to have specific environment for a project how would changing system default python3 version break that?

**Reply by candlerb** [reply]
I can't login to your system, so all I can tell you is that it's broken if you can't run nbshell.  I'm offering my best guess as to why it's broken.

**Reply by dejantep** [reply]
i just changed it back to 3.6.9 and nbshell is working as expected. 👍 Thank you for this.

**Reply by dejantep** [reply]
Does this mean when i go for netbox v3.0 and create new venv with python3.7 i need to change system default python3 to 3..7.5 aswell?

**Reply by candlerb** [reply]
No.

If you want to stick with Ubuntu 18.04, then when you install netbox v3 you'll need to do

```
PYTHON=python3.7 ./upgrade.sh
```

instead of just `./upgrade.sh` and it will build the venv with the right version of python.  (python3.8 is also available for Ubuntu 18.04, that's what I use).  The system default remains as python3.6.

Alternatively, you could migrate to Ubuntu 20.04.  This is straightforward if you're using VMs: build a new VM with v3, export the database from the old, import into the new, job done.
